libindy-crypto/src/cl/prover.rs

```diff
@@ -0,0 +1,1371 @@
+        trace!("Prover::_test_witness_credential: >>> r_claim: {:?}, r_pub_key: {:?}, r_reg: {:?}, r_cnxt_m2: {:?}", r_claim, r_pub_key, r_reg, r_cnxt_m2);
+
+        let z_calc = Pair::pair(&r_claim.witness.g_i, &r_reg.acc.acc)?
+            .mul(&Pair::pair(&r_pub_key.g, &r_claim.witness.omega)?.inverse()?)?;
+        if z_calc != r_reg.key.z {
+            return Err(IndyCryptoError::InvalidStructure("Issuer is sending incorrect data".to_string()));
+        }
+        let pair_gg_calc = Pair::pair(&r_pub_key.pk.add(&r_claim.g_i)?, &r_claim.witness.sigma_i)?;
+        let pair_gg = Pair::pair(&r_pub_key.g, &r_pub_key.g_dash)?;
+        if pair_gg_calc != pair_gg {
+            return Err(IndyCryptoError::InvalidStructure("Issuer is sending incorrect data".to_string()));
+        }
+
+        let m2 = GroupOrderElement::from_bytes(&r_cnxt_m2.to_bytes()?)?;
+
+        let pair_h1 = Pair::pair(&r_claim.sigma, &r_pub_key.y.add(&r_pub_key.h_cap.mul(&r_claim.c)?)?)?;
+        let pair_h2 = Pair::pair(
+            &r_pub_key.h0
+                .add(&r_pub_key.h1.mul(&m2)?)?
+                .add(&r_pub_key.h2.mul(&r_claim.vr_prime_prime)?)?
+                .add(&r_claim.g_i)?,
+            &r_pub_key.h_cap
+        )?;
+        if pair_h1 != pair_h2 {
+            return Err(IndyCryptoError::InvalidStructure("Issuer is sending incorrect data".to_string()));
+        }
+
+        trace!("Prover::_test_witness_credential: <<<");
+
+        Ok(())
+    }
+
+    /// Creates and returns proof builder.
+    ///
+    /// The purpose of proof builder is building of proof entity according to the given request .
+    /// # Example
+    /// ```
+    /// use indy_crypto::cl::prover::Prover;
+    /// let _proof_builder = Prover::new_proof_builder();
+    pub fn new_proof_builder() -> Result<ProofBuilder, IndyCryptoError> {
+        Ok(ProofBuilder {
+            m1_tilde: bn_rand(LARGE_M2_TILDE)?,
+            init_proofs: HashMap::new(),
+            c_list: Vec::new(),
+            tau_list: Vec::new()
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct ProofBuilder {
+    pub m1_tilde: BigNumber,
+    pub init_proofs: HashMap<String, InitProof>,
+    pub c_list: Vec<Vec<u8>>,
+    pub tau_list: Vec<Vec<u8>>,
+}
+
+impl ProofBuilder {
+    /// Add sub proof request to proof builder which will be used fo building of proof.
+    ///
+    /// # Arguments
+    /// * `proof_builder` - Proof builder.
+    /// * `key_id` - unique claim identifier.
+    /// * `claim_signature` - Claim signature.
+    /// * `claim_values` - Claim values.
+    /// * `issuer_pub_key` - Issuer public key.
+    /// * `rev_reg_pub` - (Optional) Revocation registry public.
+    /// * `sub_proof_request` -Requested attributes and predicates.
+    /// * `claim_schema` - Claim schema.
+    pub fn add_sub_proof_request(&mut self, key_id: &str, claim_signature: &ClaimSignature, claim_values: &ClaimValues, issuer_pub_key: &IssuerPublicKey,
```

Comment:
Please describe what sub-proof request is (this is part of proof request related to a particular schema-key)